daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala

```diff
@@ -0,0 +1,651 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package transaction
+
+import com.daml.lf.data.ImmArray
+import com.daml.lf.data.Ref.{
+  Location,
+  PackageId,
+  ModuleName,
+  Identifier,
+  TypeConName,
+  ChoiceName,
+  Party,
+}
+import com.daml.lf.transaction.Node.{
+  KeyWithMaintainers,
+  GenNode,
+  NodeCreate,
+  NodeFetch,
+  NodeLookupByKey,
+  NodeExercises,
+}
+import com.daml.lf.value.{Value => V}
+
+import org.scalatest.freespec.AnyFreeSpec
+
+import scala.collection.immutable.HashMap
+
+class ValidationSpec extends AnyFreeSpec {
+
+  //--[Tweaks]--
+  //
+  // A 'Tweak[X]' is a family of (small) modifications to a value of type X.
+  //
+  // This test file constructs tweaks for 'VersionedTransaction' (VTX) and classifies them
+  // as either SIGNIFICANT or INSIGNIFICANT (as reported by `isReplayedBy`).
+  //
+  // We aim to tweak every field of every ActionNode in a TX.
+  //
+  // Most fields are trivially significant; Some are trivially insignificant.  A few
+  // fields are more complicated, and have both significant and insignificant cases.
+  //
+  // The tweaks are tested by running over a hand constructed list of 'preTweakedVTXs'. We
+  // are careful to limit the combinational explosion to just what is necessary to ensure
+  // every tweak under test is hit by at least one pre-tweaked node.
+  //
+  // The testcases are organised so failure is detected and reported for a named tweak.
+
+  private class Tweak[X](val run: X => List[X])
+
+  private object Tweak {
+    def single[X](f1: X => Option[X]): Tweak[X] = {
+      def f(x: X) = f1(x) match {
+        case None => Nil
+        case Some(y) => List(y)
+      }
+      new Tweak(f)
+    }
+    def apply[X](f: X => List[X]): Tweak[X] = {
+      new Tweak(f)
+    }
+  }
```

Comment:
partial function to list seems a bit odd. Afaict we have no difference between returning an empty list and the function not being defined.

Reply:
`new Tweak(f.orElse(_ => List.empty[X]))` may work better.

Reply:
Thank you for the suggestion about `PartialFunction`. I am aware of it, but to be honest it's not something that is in my everyday scala toolbox. Also, I would be a bit unhappy to loose exhaustiveness checking for pattern matches. However, perhaps it would be idiomatic for the use case here. I wont make the change in this PR. Let's discuss offline & you can try to persuade me. :)

Reply:
I don't see any use for the `Tweak` class. A type synonym will serve equally well.

When you write a function with partial function syntax, _and_ the _expected type_ is a _non_-partial function, you will still get exhaustiveness checking. Broadly speaking, this violates the Liskov Substitution Principle (because an expression that satisfies type `t` does not satisfy type `u >: t` if warnings are errors), but not in a way that violates soundness.

So that means you can write

```scala

private type Tweak[T] = T => List[T]
private object Tweak {
  def apply[T](f: Tweak[T]): Tweak[T] = f
}

Tweak[Option[Location]] {
  case None => //blah blah
  // I get an exhaustiveness warning if I don't handle Some
}
```